### PR TITLE
Feature/add higher layer data payload flow

### PR DESCRIPTION
**Pull request by CoralMalachi**
According to section 5.10.1 in the EasyMesh Test Plan v1.0 [1],
add support for passing the Controller Certification
Higher layer data payload over 1905 trigger test flow.

Therefore, add a test in test_flows.sh to trigger the controller to
send a Higher Layer Data message to the agent and make sure 
that the message has been received in the respective agent.

#166

### CODE REVIEW

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 1]
```diff
@@ -1599,6 +1603,21 @@ bool main_thread::handle_1905_discovery_query(ieee1905_1::CmduMessageRx &cmdu_rx
...
+    LOG(DEBUG) << "Received HIGHER_LAYER_DATA_MESSAGE , mid=" << std::hex << int(mid);
+
+    // build ACK message CMDU
+    auto cmdu_tx_header = cmdu_tx.create(mid, ieee1905_1::eMessageType::ACK_MESSAGE);
+    if (!cmdu_tx_header) {
+        LOG(ERROR) << "cmdu creation of type ACK_MESSAGE, has failed";
+        return false;
+    }
+    LOG(DEBUG) << "sending ACK message back to controller";
+    return send_cmdu_to_bus(cmdu_tx, controller_bridge_mac, bridge_info.mac);
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: The message doesn't necessarily come from the controller, so you have to use the source address to identify where to send the ACK.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 2]
```diff
@@ -1599,6 +1603,21 @@ bool main_thread::handle_1905_discovery_query(ieee1905_1::CmduMessageRx &cmdu_rx
     return true;
 }
 
+bool main_thread::handle_1905_higher_layer_data_message(ieee1905_1::CmduMessageRx &cmdu_rx)
+{
+    const auto mid = cmdu_rx.getMessageId();
+    LOG(DEBUG) << "Received HIGHER_LAYER_DATA_MESSAGE , mid=" << std::hex << int(mid);
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1648]: It would be nice to also parse out the higher layer data TLV and print a bit more information about it, e.g. the protocol and the length. But not very important. (nitpick)

File: controller/src/beerocks/master/son_master_thread.cpp [hunk 3]
```diff
@@ -222,6 +222,8 @@ bool master_thread::handle_cmdu_1905_1_message(Socket *sd, ieee1905_1::CmduMessa
         return handle_cmdu_1905_channel_preference_report(sd, cmdu_rx);
     case ieee1905_1::eMessageType::CHANNEL_SELECTION_RESPONSE_MESSAGE:
         return handle_cmdu_1905_channel_selection_response(sd, cmdu_rx);
+    case ieee1905_1::eMessageType::ACK_MESSAGE:
+        return handle_cmdu_1905_ack_message(sd, cmdu_rx);
```
Comment (arnout) [inline controller/src/beerocks/master/son_master_thread.cpp]: For now this is OK, but at some point we should really handle the acks (or lack thereof) and do retransmissions. => #273

File: tools/docker/tests/test_flows.sh [hunk 4]
```diff
@@ -137,8 +137,26 @@ test_client_association() {
 }
 
 test_higher_layer_data_payload_trigger() {
-    #TODO: Implement
-    return 1
+    status "test higher layer data payload"
+    
+    mac_gateway_hex=`echo $mac_gateway | tr --delete :`
+    mac_gateway_hex="0x${mac_gateway_hex}"
```
Comment (arnout) [inline tools/docker/tests/test_flows.sh:159]: We're doing this (converting from colon-MAC to 0x) in several places now, it would make sense to make a function (in functions.sh) for it. (nitpick)

**Review by arnout — CHANGES_REQUESTED** [review]
First of all, let me commend you on the quality of this PR. It has all the features we like to see:

- Perfectly split into commits that do one thing only.
- Each commit has an extensive and correct commit message.
- The PR includes a test of the new feature.
- Everything passes CI.

There is one important fix, the destination of the ACK message. The rest of the comments are nitpicks.

Note also that this PR really isn't needed to pass 5.10.1: the only change in the controller is the handling of the ACK, and if that isn't done, the ACK just gets ignored - but this PR also just ignores it, only a bit later.

File: tools/docker/tests/test_flows.sh (hunk 4 above)
Comment (CoralMalachi) [reply tools/docker/tests/test_flows.sh:159]: you are right, I'll do it :)

**Comment by CoralMalachi** [comment]
> First of all, let me commend you on the quality of this PR. It has all the features we like to see:
> 
> * Perfectly split into commits that do one thing only.
> * Each commit has an extensive and correct commit message.
> * The PR includes a test of the new feature.
> * Everything passes CI.
> 
> There is one important fix, the destination of the ACK message. The rest of the comments are nitpicks.
> 
> Note also that this PR really isn't needed to pass 5.10.1: the only change in the controller is the handling of the ACK, and if that isn't done, the ACK just gets ignored - but this PR also just ignores it, only a bit later.

Thanks! I'll do a fixup.

**Review by arnout — APPROVED** [review]
You fixed up the wrong commit: it should be "controller:add ack handler" instead of "tools:docker:add higher layer data payload flow".

But the fix itself looks OK, so please rebase on master (making sure to fix up the right commit) so we can merge.

**Review by arnout — COMMENTED** [review]
Adding tlvHigherLevelData to the yaml + Autogenerated should be in a separate commit.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 5]
```diff
@@ -1599,6 +1603,21 @@ bool main_thread::handle_1905_discovery_query(ieee1905_1::CmduMessageRx &cmdu_rx
...
+
+    // build ACK message CMDU
+    auto cmdu_tx_header = cmdu_tx.create(mid, ieee1905_1::eMessageType::ACK_MESSAGE);
+    if (!cmdu_tx_header) {
+        LOG(ERROR) << "cmdu creation of type ACK_MESSAGE, has failed";
+        return false;
+    }
+    LOG(DEBUG) << "sending ACK message back to controller";
```
Comment (tomereli) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Please also add the MID value to the debug log

File: controller/src/beerocks/master/son_master_thread.cpp [hunk 6]
```diff
@@ -955,6 +957,13 @@ bool master_thread::handle_cmdu_1905_channel_preference_report(Socket *sd,
     return son_actions::send_cmdu_to_agent(sd, cmdu_tx);
 }
 
+bool master_thread::handle_cmdu_1905_ack_message(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx)
+{
+    auto mid = cmdu_rx.getMessageId();
+    LOG(INFO) << "Received ACK_MESSAGE, mid=" << std::hex << int(mid);
```
Comment (tomereli) [inline controller/src/beerocks/master/son_master_thread.cpp]: This should be `LOG(DEBUG)`

Comment (tomereli) [reply controller/src/beerocks/master/son_master_thread.cpp]: Should add a `//TODO` comment which mentions that the ACK should be sent to the correct task and will be done as part of agent certification (nitpick)

File: tools/docker/tests/test_flows.sh [hunk 7]
```diff
@@ -137,8 +137,26 @@ test_client_association() {
...
+    copies=200
+    
+    for i in `seq 1 $copies`
+    do
+        payload="$payload $mac_gateway_hex"
+    done
+
+    dbg "Send Higher Layer Data message"
+    eval send_bml_command '"bml_wfa_ca_controller \"DEV_SEND_1905,DestALid,$mac_agent1,MessageTypeValue,0x8018,tlv_type,0xA0,tlv_length,\
```
Comment (tomereli) [inline tools/docker/tests/test_flows.sh:171]: It will be useful if you can add a comment which explains what is being sent here as 1905 message, describing the TLVs that are part of the command.

File: tools/docker/tests/test_flows.sh [hunk 8]
```diff
@@ -137,8 +137,26 @@ test_client_association() {
...
+0x04b1,tlv_value,{0x00 $payload}\""' $redirect
+
+    dbg "Confirming higher layer data message was received"
```
Comment (tomereli) [inline tools/docker/tests/test_flows.sh]: --> dbg "Confirming higher layer data message was received in the agent"

**Review by tomereli — APPROVED** [review]
nitpick comments, but please fix them while rebasing

**Comment by CoralMalachi** [comment]
> You fixed up the wrong commit: it should be "controller:add ack handler" instead of "tools:docker:add higher layer data payload flow".
> 
> But the fix itself looks OK, so please rebase on master (making sure to fix up the right commit) so we can merge.

Ok, I'll check it now.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 5 above)
Comment (CoralMalachi) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: @tomereli  Why do we need to add the MID value to the debug log after sending the ACK? It's still the same MID.
I understood that only when we set the value of mid to zero in the `CmduMessageTx::create()` method , the value of MID is increasing by one.
is it correct?

Comment (tomereli) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: When we set MID=0, the `transport service` allocates a new MID value which is incremented by one from the last one is used.
I think that for each ACK message we should add the message type to which we are "ack'ing" and the MID value, so it will be easy to see in the log of the agent that ACK was sent, and in the controller that ACK for the same message was received.

This should have been a (nitpick) anyway, we can refactor that later on.

File: controller/src/beerocks/master/beerocks_master_mrouter.cpp [hunk 9]
```diff
@@ -56,7 +56,6 @@ bool master_mrouter::init()
         ieee1905_1::eMessageType::CHANNEL_PREFERENCE_REPORT_MESSAGE,
         ieee1905_1::eMessageType::CHANNEL_SELECTION_RESPONSE_MESSAGE,
         ieee1905_1::eMessageType::OPERATING_CHANNEL_REPORT_MESSAGE,
-        ieee1905_1::eMessageType::TOPOLOGY_DISCOVERY_MESSAGE,
```
Comment (tomereli) [inline controller/src/beerocks/master/beerocks_master_mrouter.cpp]: Why is this removed??

**Review by tomereli — CHANGES_REQUESTED** [review]

**Review by tomereli — APPROVED** [review]

File: tools/docker/tests/test_flows.sh (hunk 4 above)
Comment (CoralMalachi) [reply tools/docker/tests/test_flows.sh:159]: #282
